storage/backend/null/null.go

```diff
@@ -1,104 +1,162 @@
 package null
 
 import (
+	"os"
 	"path/filepath"
+	"runtime"
+	"strings"
 	"time"
 
+	log "github.com/Sirupsen/logrus"
+	"github.com/contiv/errored"
 	"github.com/contiv/volplugin/storage"
 )
 
 // BackendName is string for no-op storage backend
 const BackendName = "null"
 
+// gNullDriver is the singleton driver instace to keep all state in memory
+var gNullDriver *Driver
+
+func getFunctionName() string {
+	if pc, _, _, ok := runtime.Caller(1); ok {
+		f := strings.Split(runtime.FuncForPC(pc).Name(), "/")
+		return f[len(f)-1]
+	}
+	return "unknown"
+}
+
 // Driver implements a no-op storage driver for volplugin.
 //
 // This is intended for regressing volplugin
 type Driver struct {
 	mountpath string
+	created   map[string]bool
+	mounted   map[string]bool
 }
 
 // NewDriver is a generator for Driver structs. It is used by the storage
 // framework to yield new drivers on every creation.
 func NewDriver(mountpath string) storage.Driver {
-	return &Driver{mountpath: mountpath}
+	log.Infof("In %s", getFunctionName())
+	if gNullDriver != nil {
+		return gNullDriver
+	}
+	gNullDriver = &Driver{
+		mountpath: mountpath,
+		created:   make(map[string]bool),
+		mounted:   make(map[string]bool),
+	}
+	return gNullDriver
 }
 
 // Name returns the null backend string
 func (d *Driver) Name() string {
+	log.Infof("In %q", getFunctionName())
 	return BackendName
 }
 
-// Create is a noop.
-func (d *Driver) Create(storage.DriverOptions) error {
+// Create records created volume.
+func (d *Driver) Create(do storage.DriverOptions) error {
+	log.Infof("In %q", getFunctionName())
+	d.created[filepath.Join(d.mountpath, do.Volume.Params["pool"], do.Volume.Name)] = true
 	return nil
 }
 
 // Format is a noop.
 func (d *Driver) Format(do storage.DriverOptions) error {
+	log.Infof("In %q", getFunctionName())
 	return nil
 }
 
-// Destroy is a noop.
+// Destroy deletes volume entry
 func (d *Driver) Destroy(do storage.DriverOptions) error {
+	log.Infof("In %q", getFunctionName())
+	delete(d.created, filepath.Join(d.mountpath, do.Volume.Params["pool"], do.Volume.Name))
 	return nil
 }
 
 // List Volumes returns empty list.
 func (d *Driver) List(lo storage.ListOptions) ([]storage.Volume, error) {
+	log.Infof("In %q", getFunctionName())
 	return []storage.Volume{}, nil
 }
 
-// Mount is a noop
+// Mount records mounted volume
 func (d *Driver) Mount(do storage.DriverOptions) (*storage.Mount, error) {
+	log.Infof("In %q", getFunctionName())
+	if err := os.MkdirAll(d.mountpath, 0700); err != nil && !os.IsExist(err) {
```

Comment:
> I really think this shouldn't write to the filesystem. Perhaps we should instead map the paths and check for path existence instead, and return a useful error. 

@erikh just to understand, do you mean we should not create the `d.mountpath` and `volumePath` directories at all? OR just create them once and avoid subsequent `os.MkdirAll()` calls.

IIRC we do need to create these directories otherwise `docker run` complains. I think I can optimize by checking for existence of the directory (by keeping local state) and not call `os.MkdirAll()` at all. Does that work?

> Also, I notice these paths are not getting cleaned up in Unmount.

yeah, that was part intentional and part laziness. Basically by not having an explicit `rmdir` call on every unmount I saved some work (both mine and system's :) ) of not making the extra system call. But thinking more adding the extra call might be better as it cleans up state after a successful test. Or we can do it just once like we do in `purgeVolumes()` for ceph. Latter might be better, wdyt?